CommunityToolkit.WinUI/Extensions/DispatcherQueueExtensions{T}.cs

```diff
@@ -0,0 +1,148 @@
+        /// </summary>
+        /// <typeparam name="T1">The type of the first state to capture.</typeparam>
+        /// <typeparam name="T2">The type of the second state to capture.</typeparam>
+        /// <param name="dispatcherQueue">The target <see cref="DispatcherQueue"/> to invoke the code on.</param>
+        /// <param name="priority"> The desired priority for the callback to schedule.</param>
+        /// <param name="callback">The input <see cref="DispatcherQueueHandler{T}"/> callback to enqueue.</param>
+        /// <param name="state1">The first input state to capture and pass to the callback.</param>
+        /// <param name="state2">The second input state to capture and pass to the callback.</param>
+        /// <returns>Whether or not the task was added to the queue.</returns>
+        /// <exception cref="Exception">Thrown when the enqueue operation fails.</exception>
+        public static unsafe bool TryEnqueue<T1, T2>(this DispatcherQueue dispatcherQueue, DispatcherQueuePriority priority, DispatcherQueueHandler<T1, T2> callback, T1 state1, T2 state2)
+            where T1 : class
+            where T2 : class
+        {
+            return TryEnqueue(dispatcherQueue, priority, DispatcherQueueProxyHandler2.Create(callback, state1, state2));
+        }
+
+        /// <summary>
+        /// Adds a task to the <see cref="DispatcherQueue"/> which will be executed on the thread associated with it.
+        /// </summary>
+        /// <param name="dispatcherQueue">The target <see cref="DispatcherQueue"/> to invoke the code on.</param>
+        /// <param name="priority"> The desired priority for the callback to schedule (if available).</param>
+        /// <param name="dispatcherQueueHandler">The input callback to enqueue.</param>
+        /// <returns>Whether or not the task was added to the queue.</returns>
+        /// <exception cref="Exception">Thrown when the enqueue operation fails.</exception>
+        private static unsafe bool TryEnqueue<THandler>(DispatcherQueue dispatcherQueue, DispatcherQueuePriority? priority, THandler* dispatcherQueueHandler)
+            where THandler : unmanaged, IDispatcherQueueHandler
+        {
+            bool success;
+            int hResult;
+
+            try
+            {
+                IDispatcherQueue* dispatcherQueuePtr = (IDispatcherQueue*)((IWinRTObject)dispatcherQueue).NativeObject.ThisPtr;
+
+                if (priority.HasValue)
+                {
+                    hResult = dispatcherQueuePtr->TryEnqueueWithPriority(priority.GetValueOrDefault(), dispatcherQueueHandler, (byte*)&success);
+                }
+                else
+                {
+                    hResult = dispatcherQueuePtr->TryEnqueue(dispatcherQueueHandler, (byte*)&success);
+                }
+
+                GC.KeepAlive(dispatcherQueue);
+            }
+            finally
+            {
+                dispatcherQueueHandler->Release();
```

Comment:
I don't see a corresponding `->Add()` call for this, am I missing something?

Reply:
I just skipped that and initialized the ref count to 1 directly in the static constructor methods 😄

https://github.com/windows-toolkit/WindowsCommunityToolkit/blob/972402df917de3ce351f6619b670b7d140a04866/CommunityToolkit.WinUI/Extensions/Interop/DispatcherQueueProxyHandler1.cs#L78-L83

Reply:
Ah. I'd recommend moving that ->Release call near the allocation site then, to make it clear that this is removing the initial reference.

Reply:
I did this to reduce code duplication and avoid having to put a `try/finally` block in each one of the public APIs here, not really seeing a way around this without either going back to that or adding some `ComPtr<T>` type or something (and still end up with more code than we have right now though). I can definitely document this better so it's clearer for anyone else reading that line though, good catch! 🙂